lib/nmatrix/nmatrix.rb

```diff
@@ -90,6 +90,62 @@ def size(shape)
       shape = [shape,shape] unless shape.is_a?(Array)
       (0...shape.size).inject(1) { |x,i| x * shape[i] }
     end
+
+    # Make N-D coordinate arrays for vectorized evaluations of
+    # N-D scalar/vector fields over N-D grids, given N one-dimensional
+    # coordinate arrays arrs. N > 1.
+    #
+    # call-seq:
+    #     meshgrid(arrs) -> Array of NMatrix
+    #     meshgrid(arrs, options) -> Array of NMatrix
+    #
+    # * *Arguments* :
+    #   - +arrs+ -> Array of N one-dimensional coordinate arrays (Array or NMatrix)
+    #   - +options+ -> Hash with options (:sparse and :ij both Boolean, false by default)
+    # * *Returns* :
+    #   - Array of N N-D NMatrixes
+    # * *Examples* :
+    #     x, y = NMatrix::meshgrid([[1, 2, 3], [4, 5]])
+    #     x.to_a #<= [[1, 2, 3], [1, 2, 3]]
+    #     y.to_a #<= [[4, 4, 4], [5, 5, 5]]
+    #
+    # * *Using* *options* :
+    #
+    #     x, y = NMatrix::meshgrid([[1, 2, 3], [4, 5]], sparse: true)
+    #     x.to_a #<= [[1, 2, 3]]
+    #     y.to_a #<= [[4], [5]]
+    #
+    #     x, y = NMatrix::meshgrid([[1, 2, 3], [4, 5]], ij: true)
+    #     x.to_a #<= [[1, 1], [2, 2], [3, 3]]
+    #     y.to_a #<= [[4, 5], [4, 5], [4, 5]]
+    def meshgrid(arrs, options = {})
+      raise(ArgumentError, 'Expected at least 2 arrays.') if arrs.size < 2
+      mats = arrs.map do |arr|
+        case arr
+        when NMatrix
+          arr
+          raise(ArgumentError, 'Expected one-dimensional arrays.') if arr.effective_dim != 1
+        when Array
+          NMatrix.new([arr.size], arr)
+        else
+          raise(ArgumentError, 'Expected one-dimensional arrays.')
+        end
+      end
+      mats[0], mats[1] = mats[1], mats[0] unless options[:ij]
+      new_dim = mats.size
+      lengths = mats.map(&:size)
+      result = mats.map.with_index do |matrix, axis|
+        new_shape = Array.new(new_dim, 1)
+        new_shape[axis] = lengths[axis]
+        new_arr = matrix.reshape(new_shape)
+        new_dim.times do |i|
+          new_arr = new_arr.repeat(lengths[i], i) if axis != i
+        end unless options[:sparse]
```

Comment:
The thing with `#repeat` is that its trying to increase the elements present in the NMatrix.

NMatrix is a library for static storage of data, i.e data that does not change in size frequently. So we should encourage users to not use too many methods that move away from this philosophy.

`repeat` is a handy method to have in case you need it, but its use should be minimal, especially in the nmatrix code base.

You're really not reinventing the wheel with Arrays. You'll probably end up using some thing like `Array#<<` instead. Since Arrays are optimized for dynamically increasing and decreasing the data stored in them, it will be much more efficient. You can just convert the Array to NMatrix at the end of the loop, so `result` will store the NMatrix objects directly.

Discouraging chaining is non Ruby-esque, if you know what I mean.